### PR TITLE
OS2FORMS-372 Added Easy person validation by CPR and name

### PR DESCRIPTION
**Pull request by andriyun**
Changes in PR implement a composite field "Easy CPR person lookup"

With default settings you will get  2 fields:
- CPR number
- Name

Value from CPR number validates over correct CPR number.
For valid CPR, Name it matches with data retrieved from Service platformen.

### CODE REVIEW

**Comment by andriyun** [comment]
Hi @rimi-itk 
I've sent you an invite for collaborating on this repository. When you get it accepted, I can assign you for review on this PR

File: src/Plugin/WebformElement/Os2formsPersonLookup.php [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+   * Form API callback. Make sure CPR is valid and belong provided name.
+   */
+  public static function validatePerson(array &$element, FormStateInterface $form_state, array &$completed_form) {
+    $values = $element['#value'];
+    $cpr_number = str_replace('-', '', $values['cpr_number']);
+    $os2web_datalookup_plugins = \Drupal::service('plugin.manager.os2web_datalookup');
+
+    /** @var \Drupal\os2web_datalookup\Plugin\os2web\DataLookup\ServiceplatformenCPR $servicePlatformentCprPlugin */
+    $servicePlatformentCprPlugin = $os2web_datalookup_plugins->createInstance('serviceplatformen_cpr');
+    $peronsData = $servicePlatformentCprPlugin->cprBasicInformation($cpr_number);
```
Comment (rimi-itk) [inline src/Plugin/WebformElement/Os2formsPersonLookup.php:104]: Typo in variable name; it should be `personData`.

File: src/Plugin/WebformElement/Os2formsPersonLookup.php [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+   * {@inheritdoc}
+   */
+  public function prepare(array &$element, WebformSubmissionInterface $webform_submission = NULL) {
+    parent::prepare($element, $webform_submission);
+  }
+
+  /**
+   * Form API callback. Make sure CPR is valid and belong provided name.
+   */
+  public static function validatePerson(array &$element, FormStateInterface $form_state, array &$completed_form) {
```
Comment (rimi-itk) [inline src/Plugin/WebformElement/Os2formsPersonLookup.php:97]: Why is this a static function? Is it a requirement by Drupal or Webforms?

File: src/Plugin/WebformElement/Os2formsPersonLookup.php [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    $os2web_datalookup_plugins = \Drupal::service('plugin.manager.os2web_datalookup');
+
+    /** @var \Drupal\os2web_datalookup\Plugin\os2web\DataLookup\ServiceplatformenCPR $servicePlatformentCprPlugin */
+    $servicePlatformentCprPlugin = $os2web_datalookup_plugins->createInstance('serviceplatformen_cpr');
+    $peronsData = $servicePlatformentCprPlugin->cprBasicInformation($cpr_number);
+
+    if (!$peronsData['status']) {
+      $error = isset($peronsData['text']) ? $peronsData['text'] : (isset($peronsData['error']) ? $peronsData['error'] : t('Can not verify CPR Number'));
+      \Drupal::logger('os2forms')->warning(t('os2forms_person_lookup - data lookup error: @error', ['@error' => $error]));
+      $form_state->setError($element['cpr_number'], t('CPR is not valid.'));
```
Comment (rimi-itk) [inline src/Plugin/WebformElement/Os2formsPersonLookup.php:109]: You should `use StringTranslationTrait` in the class and use `$this->t` here (cf. <https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21StringTranslation%21StringTranslationTrait.php/trait/StringTranslationTrait/9>).

File: src/Plugin/WebformElement/Os2formsPersonLookup.php [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    parent::prepare($element, $webform_submission);
+  }
+
+  /**
+   * Form API callback. Make sure CPR is valid and belong provided name.
+   */
+  public static function validatePerson(array &$element, FormStateInterface $form_state, array &$completed_form) {
+    $values = $element['#value'];
+    $cpr_number = str_replace('-', '', $values['cpr_number']);
+    $os2web_datalookup_plugins = \Drupal::service('plugin.manager.os2web_datalookup');
```
Comment (rimi-itk) [inline src/Plugin/WebformElement/Os2formsPersonLookup.php:100]: If possible, you should use dependency injection (cf. https://www.drupal.org/docs/drupal-apis/services-and-dependency-injection/dependency-injection-for-a-form) to use this service.

**Review by rimi-itk — CHANGES_REQUESTED** [review]
Looks good. Some minor issues and suggestions should be addressed.

File: src/Plugin/WebformElement/Os2formsPersonLookup.php (hunk 1 above)
Comment (andriyun) [reply src/Plugin/WebformElement/Os2formsPersonLookup.php]: Fixed

File: src/Plugin/WebformElement/Os2formsPersonLookup.php (hunk 2 above)
Comment (andriyun) [reply src/Plugin/WebformElement/Os2formsPersonLookup.php:97]: It's a form element validating callback.
This kind of callback is called via `call_user_func_array` function
https://git.drupalcode.org/project/drupal/-/blob/9.4.x/core/lib/Drupal/Core/Form/FormValidator.php#L282

In this case, a function is called without context and should be declared as static.

File: src/Plugin/WebformElement/Os2formsPersonLookup.php (hunk 3 above)
Comment (andriyun) [reply src/Plugin/WebformElement/Os2formsPersonLookup.php:109]: It's not possible to use `$this` context in static functions

File: src/Plugin/WebformElement/Os2formsPersonLookup.php (hunk 2 above)
Comment (rimi-itk) [reply src/Plugin/WebformElement/Os2formsPersonLookup.php:97]: I would expect it to work non-statically if we use

```php
$element['#element_validate'][] = [$this, 'validatePerson'];
```

in `prepareElementValidateCallbacks`, but it seems that the Webform way to do it is statically. We'll keep it static.

**Review by rimi-itk — APPROVED** [review]

File: src/Plugin/WebformElement/Os2formsPersonLookup.php [hunk 5]
```diff
@@ -116,7 +117,8 @@ public static function validatePerson(array &$element, FormStateInterface $form_
       return;
     }
 
-    if ($personsData['adresseringsnavn'] != $values['name']) {
+    $helper = new NameHelper();
+    if ($helper->compareNames($personsData['adresseringsnavn'], $values['name']) != 0) {
```
Comment (rimi-itk) [inline src/Plugin/WebformElement/Os2formsPersonLookup.php:121]: You may want to use `!==` here.

**Review by rimi-itk — APPROVED** [review]
Approved with a single comment on `!=` vs. `!==`.

**Comment by andriyun** [comment]
image

There should be the same error message: Navn og CPR nummer stemmer ikke overens

**Comment by andriyun** [comment]
Add danish translation for English strings and element labels

**Comment by andriyun** [comment]
Easy CPR person lookup -> CPR / Navn validering

**Comment by andriyun** [comment]
Hi @rimi-itk 
I've added changes related to the notes I took during the last meeting. Could you review them?
Thx

File: src/Plugin/WebformElement/Os2formsPersonLookup.php (hunk 5 above)
Comment (andriyun) [reply src/Plugin/WebformElement/Os2formsPersonLookup.php]: Solved

File: src/Element/Os2formsPersonLookup.php [hunk 6]
```diff
@@ -18,12 +18,12 @@ public static function getCompositeElements(array $element) {
     $elements = [];
     $elements['cpr_number'] = [
       '#type' => 'textfield',
-      '#title' => t('CPR number'),
+      '#title' => t('CPR nummer'),
```
Comment (rimi-itk) [inline src/Element/Os2formsPersonLookup.php:21]: In Danish this is spelled with a dash, i.e. “CPR-nummer” (many native Danish writers don't know this).

**Review by rimi-itk — CHANGES_REQUESTED** [review]
Danish is a tricky language. A typo should be fixed (in multiple places).

File: src/Element/Os2formsPersonLookup.php (hunk 6 above)
Comment (andriyun) [reply src/Element/Os2formsPersonLookup.php]: Fixed

**Review by rimi-itk — APPROVED** [review]

**Pull request merged**